src/main/java/org/usfirst/frc4048/common/apriltags/TCPServer.java

```diff
@@ -0,0 +1,113 @@
+package org.usfirst.frc4048.common.apriltags;
+
+import edu.wpi.first.wpilibj.DriverStation;
+
+import java.io.BufferedInputStream;
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.util.LinkedList;
+import java.util.Queue;
+import java.util.concurrent.LinkedTransferQueue;
+
+public abstract class TCPServer<T> extends Thread {
+
+    private final int port;
+    private ServerSocket serverSocket;
+    private Socket clientSocket;
+    private DataInputStream dataInputStream;
+    private BufferedInputStream bufferedInputStream;
+    private boolean running = false;
+    private final LinkedTransferQueue<T> readings = new LinkedTransferQueue<>();
+
+    /**
+     * @param port for network communication. Port must match between client and server (range 5800-5810 (rio limitations))
+     */
+    public TCPServer(int port) {
+        this.port = port;
+    }
+
+
+    @Override
+    public void run() {
+        try {
+            this.serverSocket = new ServerSocket(port);
+            this.clientSocket = null;
+        } catch (IOException e2) {
+            DriverStation.reportError("Could not start server on port " + port, true);
+            this.serverSocket = null;
+        }
+        running = true;
+        while (running) {
+            if (serverSocket == null) {
+                escape();
+                return;
+            }
+            if (this.clientSocket == null) {
+                clientSocket = waitForClient();
```

Comment:
This would try to create the connection for ever - maybe do some of:
- introduce a delay between attempts
- limit the number of attempts (quit after a while)
- Add timeout to the `accept` call.
The current behavior is for the accept to wait forever, with the only option to stop it in case there is no client is by `interrupt` ing the thread. Adding timeout (I saw this:
```
  try {    
      myServer = new ServerSocket(port)
      myServer.setSoTimeout(2000); 
      //YOU MUST DO THIS ANYTIME TO ASSIGN new ServerSocket() to myServer‼!
      clientSocket = myServer.accept();
      //In this case, after 2 seconds the below interruption will be thrown
  }
```
) - may make this more ropbust in case of errors

Reply:
A couple of thoughts
- I like the idea of a delay between attempts. I have one possible concern (depending on the implementation of the client connection code). If the client also has a delay in attempts, it may request while the server thread is waiting. Then, when the server thread is ready, the client may be waiting between attempts. I do not know where the client code is, so I don't know if it is a problem, but I thought I would bring it up.
- I understand that limiting the number of attempts would solve the problem of lets say the PI not being connected and then the robot always listening for a connection that will never come. However, the PI and robot are not always started at the same time. There were a couple of times last year where someone forgot to plug in the PI, followed by a field delay, and then someone notices that the PI is not plugged in and connects it before the match. If we implemented a max number of attempts, then the PI would not connect to the server. A solution to this would be to restart the robot code on driver station (however that button only works 30% of time for some reason) and the drive team (if aware of the need for a restart to connect the PI) may not have time to reboot the RIO.
- _**IF**_ we don't have a delay between attempts, is it still a good idea to have a timeout? Should it be long or short?

Reply:
In general, I don't think waiting on a socket to connect is resource consuming, so looping forever is OK. The benefit of having the Pi connect whenever we plug it in is a good one.
Having a timed `accept` that stops after a while would allow us to do things like stopping the loop if the robot stops. I think that is necessary but we have to test it.
As long as the `accept` has a timeout of a few seconds and the delay between iterations is much shorter (100ms?) then I don't think there will be an issue.

Reply:
done